integrations/helm/release/utils.go

```diff
@@ -0,0 +1,131 @@
+package release
+
+import (
+	"encoding/json"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/go-kit/kit/log"
+	ifv1 "github.com/weaveworks/flux/apis/helm.integrations.flux.weave.works/v1alpha"
+)
+
+// collectValues ... assembles overriding parameters and outputs a
+// map suitable for passing to helm client API.
+// Parameters with names containing "." will result in nested maps.
+// Maps from all parameters are merged together, with a possible overwriting
+// if a parameter is provided twice with different values.
+func collectValues(logger log.Logger, params []ifv1.HelmChartParam) (map[string]interface{}, error) {
+	base := map[string]interface{}{}
+	if params == nil || len(params) == 0 {
+		return base, nil
+	}
+
+	var vu interface{}
+	var err error
+	regx := regexp.MustCompile(`^\[.*\]$`)
+
+	for _, p := range params {
+		k, v := cleanup(p.Name, p.Value)
+		if k == "" {
+			continue
+		}
+
+		vu = v
+		if match := regx.Match([]byte(v)); match {
+			vu, err = unwrap(v)
+			if err != nil {
+				return nil, err
+			}
+		}
+		pMap := mappifyValueOverride(k, vu)
+		base = mergeValues(base, pMap)
+	}
+
+	logger.Log("debug", fmt.Sprintf("override parameters in a data structure: %#v", base))
+
+	return base, nil
+}
+
+func cleanup(k, v string) (string, string) {
+	k = strings.TrimSpace(k)
+	k = strings.Trim(k, "\n")
+
+	v = strings.TrimSpace(v)
+	v = strings.Trim(v, "\n")
+
+	return k, v
+}
+
+func reverse(s []string) []string {
+	for i, j := 0, len(s)-1; i < j; i, j = i+1, j-1 {
+		s[i], s[j] = s[j], s[i]
+	}
+	return s
+}
+
+// mappifyValueOverride ... takes a parameter and its value, and creates
+// a corresponding map suitable for passing to helm client API
+// to override default values
+func mappifyValueOverride(k string, v interface{}) map[string]interface{} {
+	nests := reverse(strings.Split(k, "."))
+
+	inner := map[string]interface{}{}
+	outer := map[string]interface{}{}
+	for i, n := range nests {
+		switch i {
+		case 0:
+			inner[n] = v
+		default:
+			outer = map[string]interface{}{
+				nests[i]: inner,
+			}
+			inner = outer
+		}
+
+	}
+	return inner
+}
+
+// mergeValues ... merges two, possibly nested, maps
+// (copied from kubernetes/helm/cmd/helm/install.go)
+func mergeValues(dest map[string]interface{}, src map[string]interface{}) map[string]interface{} {
+	for k, v := range src {
+		// If the key doesn't exist already, then just set the key to that value
+		if _, exists := dest[k]; !exists {
+			dest[k] = v
+			continue
+		}
+		nextMap, ok := v.(map[string]interface{})
+		// If it isn't another map, overwrite the value
+		if !ok {
+			dest[k] = v
+			continue
+		}
+		// If the key doesn't exist already, then just set the key to that value
+		if _, exists := dest[k]; !exists {
```

Comment:
This check is the same as the one on line 95 and will never be true.
The lines 99 to 104 don't modify `dest` or `k` so anything caught here would be caught by line 95.
Is this correct?

Reply:
This method was copied from kubernetes/helm/cmd/helm/install.go, so, to tell the truth, I did not spend much time looking at it.
You are right about your observation from what I can see. I shall write a test for this method to confirm it behaves.

Reply:
Actually, need to think about it more. It's good to think ....

Reply:
Removed the duplicate, renamed the function and added tests. The tests confirm you were right, that the functionality truly does not change whether or not those lines are there.